src/lib/formats/rpk.cpp

```diff
@@ -0,0 +1,459 @@
+	}
+
+	// determine the type of RAM
+	rpk_socket::socket_type type;
+	std::string const *const ram_type = ram_resource_node.get_attribute_string_ptr("type");
+	if (ram_type && *ram_type == "persistent")
+		type = rpk_socket::socket_type::PERSISTENT_RAM;
+	else
+		type = rpk_socket::socket_type::RAM;
+
+	// persistent RAM needs a file name
+	std::string file;
+	if (type == rpk_socket::socket_type::PERSISTENT_RAM)
+	{
+		std::string const *const ram_filename = ram_resource_node.get_attribute_string_ptr("file");
+		if (ram_filename == nullptr)
+			return rpk_reader::error::INVALID_RAM_SPEC; // <ram type='persistent'> must have a 'file' attribute
+		file = *ram_filename;
+	}
+
+	// finally add the socket
+	m_sockets.emplace_back(*this, std::move(id), type, std::move(file), std::nullopt, length);
+	return std::error_condition();
+}
+
+
+/***************************************************************************
+	RPK SOCKET
+***************************************************************************/
+
+//-------------------------------------------------
+//  ctor
+//-------------------------------------------------
+
+util::rpk_socket::rpk_socket(rpk_file &rpk, std::string &&id, socket_type type, std::string &&filename, std::optional<hash_collection> &&hashes, std::optional<std::uint32_t> length)
+	: m_rpk(rpk)
+	, m_id(id)
+	, m_type(type)
+	, m_filename(filename)
+	, m_hashes(std::move(hashes))
+	, m_length(length)
+{
+}
+
+
+//-------------------------------------------------
+//  dtor
+//-------------------------------------------------
+
+util::rpk_socket::~rpk_socket()
+{
+}
+
+
+//-------------------------------------------------
+//  read_file
+//-------------------------------------------------
+
+std::error_condition util::rpk_socket::read_file(std::vector<std::uint8_t> &result) const
+{
+	// find the file
+	if (m_rpk.zipfile().search(m_filename, false) < 0)
+		return rpk_reader::error::INVALID_FILE_REF;
+
+	// prepare a buffer
+	result.resize(m_rpk.zipfile().current_uncompressed_length());
+
+	// read the file
+	std::error_condition const ziperr = m_rpk.zipfile().decompress(&result[0], m_rpk.zipfile().current_uncompressed_length());
+	if (ziperr)
+		return ziperr;
+
+	// perform hash checks, if appropriate
+	if (m_hashes.has_value())
+	{
+		util::hash_collection actual_hashes;
+		actual_hashes.compute(&result[0], result.size(), m_hashes->hash_types().c_str());
+
+		if (actual_hashes != m_hashes)
+			return rpk_reader::error::INVALID_FILE_REF; // Hash check failed
+	}
+
+	// success!
+	return std::error_condition();
+}
+
+
+/***************************************************************************
+	RPK EXCEPTION HANDLING
+***************************************************************************/
+
+//-------------------------------------------------
+//  rpk_category - gets the RPK error category instance
+//-------------------------------------------------
+
+std::error_category const &util::rpk_category() noexcept
```

Comment:
Wouldn’t this give an extra qualifier error?  The `rpk_category` should be in the same namespace as the error enum it uses anyway.